harvester/network-controller-harvester
# Solution two issues related to node change

**Problem:**

1. Clear link statuses of the deleted node in all the link monitor
2. Skip vlanconfig webhook(includes validator and mutator) if the spec is not changed.

**Solution:**

1. Add `OnRemove` function in the node controller to clear link statuses
2. Use reflect to compare the spec of the old vlanconfig and new vlanconfig. If it's not changed, skip the webhook.

By the way, use a more elegant way to upgrade golang.org/x/net 

**Related Issue:**
https://github.com/harvester/harvester/issues/3685
https://github.com/harvester/harvester/issues/3697

**Test plan:**

- Set up a cluster with one node which has more than one NIC for the custom cluster network.
- Add a new cluster network and network config which selects all nodes.
- Join a new node whose NICs are same with the first node. It would be added successfully and the network config also takes effect. The issue https://github.com/harvester/harvester/issues/3685 is validated.
- Execute `kubectl get lms -o yaml` to check link statuses of all nodes.
- Delete the second node. Check link statuses again and the link statuses of the removed node are deleted.  The issue https://github.com/harvester/harvester/issues/3697 is validated.
